client/db/src/lib.rs

```diff
@@ -713,22 +713,24 @@ pub(crate) struct FrozenForDuration<T: Clone> {
 }
 
 impl<T: Clone> FrozenForDuration<T> {
-	fn new(duration: std::time::Duration, initial: T) -> Self {
+	fn new(duration: std::time::Duration) -> Self {
 		Self {
 			duration,
-			value: Frozen { at: std::time::Instant::now(), value: initial }.into(),
+			value: Frozen { at: std::time::Instant::now(), value: None }.into(),
 		}
 	}
 
 	fn take_or_else<F>(&self, f: F) -> T where F: FnOnce() -> T {
-		if self.value.read().at.elapsed() > self.duration {
-			let mut write_lock = self.value.write();
+		let read_lock = self.value.upgradable_read();
```

Comment:
Upgradeable read locks are exclusive as well. So there could exists readers at the same time, but no concurrent other upgradable read lock. So, please just use a write lock directly.

Reply:
Right, I'll just replace with mutex then